### ISSUE TITLE
currently in use by a listener or a rule

### ISSUE THREAD

**Issue by Zogoo** [hashicorp/terraform-provider-aws#16889]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform CLI and Terraform AWS Provider Version
terraform 12.0.6
provider.aws: version = "~> 3.21"

### Affected Resource(s)

aws_lb_target_group

* aws_XXXXX

### Terraform Configuration Files

```terraform
  # module.load_balancer.aws_lb_target_group.target_group (deposed object ****) will be destroyed
  - resource "aws_lb_target_group" "targe_group" {
 ```

### Debug Output

When target group which used for ECS service get renewed Terraform requires that delete existing one, but Terraform fails and says that target group assigned to LoadBalancer.

### Panic Output

```terraform
Error: Error deleting Target Group: ResourceInUse: Target group 'arn:aws:elasticloadbalancing:ap-northeast-1:****:targetgroup/target_group_name/*****' is currently in use by a listener or a rule
	status code: 400, request id: 67183cbe-eb17-41af-9d83-e68028eec01c
```

### Expected Behavior

Terraform needs to remove dependency that assigned into LoadBalancer listener group before delete Target group object.

### Actual Behavior

It fails

### Steps to Reproduce

1. `terraform apply`

### Important Factoids

Nothing

**Comment by nikolay-te** [comment]
Maybe the workaround from here might help https://github.com/hashicorp/terraform-provider-aws/issues/1315

**Comment by Zogoo** [comment]
> Maybe the workaround from here might help #1315

Thanks for your suggestion.
 
Actually I have already set lifecycle in my lb target group resources.

```terraform
 lifecycle {
    create_before_destroy = true
  }
```

Problem was if you use LB for ECS service dependencies are not deleted by terraform

**Comment by CDRayn** [comment]
I've encountered a similar problem. The target group that the listener is referring to, is used by an ECS service running Fargate tasks. One fix is to manually delete the listener after the `terraform apply` fails and then run `terraform apply` again. While not ideal, this seemed to resolve the problem.

**Comment by MartinEmrich** [comment]
similar issue here. I have an `aws_lb_target_group`, and an `aws_lb_listener` referring to it.
if the target group needs to be replaced, terraform fails, instead of deleting the listener before, and recreating it after replacing the target group.

create_before_destroy is no option, as both the target group name have to be unique.

"workaround" is to manually delete the listener(s) before `terraform apply`. TF will detect their absence and recreate them after replacing the corresponding target group(s).

**Comment by tim-mardesen** [comment]
@MartinEmrich I wonder if tainting those `aws_lb_listener_rule` would allow for terraform to destroy those first? Also wondering if there is a way in which we can force terraform to reorder the changes it is about to perform. If that were the case, we could atleast code around this with dependency management

**Comment by clint-truss** [comment]
@tim-mardesen I tried using `depends_on` both in the aws_lb_listener as well as in the aws_lb_target_group to see if I could help terraform understand which order to use here. It did not change the behavior. I'm running terraform 0.12.31 and AWS provider 3.61

**Comment by rnsc** [comment]
> similar issue here. I have an `aws_lb_target_group`, and an `aws_lb_listener` referring to it. if the target group needs to be replaced, terraform fails, instead of deleting the listener before, and recreating it after replacing the target group.
> 
> create_before_destroy is no option, as both the target group name have to be unique.
> 
> "workaround" is to manually delete the listener(s) before `terraform apply`. TF will detect their absence and recreate them after replacing the corresponding target group(s).

Exactly the same issue for me today.
Had to rename my target groups and Terraform failed to free up the listener first.

**Comment by jbg** [comment]
> create_before_destroy is no option, as both the target group name have to be unique.

Use `create_before_destroy` along with `name_prefix` so that the new one gets a new name.

**Comment by MartinEmrich** [comment]
@jbg If non-terraform-controlled things (ECS, Kubernetes ressources, ...) depend on a predefined name, that won't work: with `name_prefix`, and without `name`, terraform will append a random string to the `name_prefix` value, so the final target group name will change, and won't be fixed and deterministic.

**Comment by jbg** [comment]
yup, any time you have dependencies that are unknown to Terraform you will run into this kind of trouble. The obvious solution is to import those other resources into TF rather than rely on error-prone manual handling of dependencies.

**Comment by taylorsilva** [comment]
> > create_before_destroy is no option, as both the target group name have to be unique.
> 
> Use `create_before_destroy` along with `name_prefix` so that the new one gets a new name.

This worked for me. Changed my `aws_lb_target_group` to this:

```diff
resource "aws_lb_target_group" "test" {
-  name = "test"
+  name_prefix = "test"
...

+  lifecycle {
+    create_before_destroy = true
+  }
}
```

Plan applied successfully after that.

**Comment by DanielViglione** [comment]
> I've encountered a similar problem. The target group that the listener is referring to, is used by an ECS service running Fargate tasks. One fix is to manually delete the listener after the `terraform apply` fails and then run `terraform apply` again. While not ideal, this seemed to resolve the problem.

You shouldn't have to manually delete anything. That's the whole point of using IAC. To use Code, no manual work.

**Comment by jbg** [comment]
> You shouldn't have to manually delete anything. That's the whole point of using IAC. To use Code, no manual work.

That assumes you used IAC properly, and didn't create resources manually (or using other automation) outside of the IAC's state that it doesn't know about.

**Comment by DanielViglione** [comment]
> > You shouldn't have to manually delete anything. That's the whole point of using IAC. To use Code, no manual work.
> 
> That assumes you used IAC properly, and didn't create resources manually (or using other automation) outside of the IAC's state that it doesn't know about.

It goes without say that nothing was done manually.

**Comment by jbg** [comment]
If you read the comments above yours, many of the people landing in this issue have dependencies from outside Terraform which is the cause of their issue. Of course, I don't know anything about your situation since you didn't share anything about your situation.

**Comment by DanielViglione** [comment]
I explicitly quoted @ CDRayn . He didn't cite any prior manual work. He explicitly cited the interdependency of ECS, Fargate Tasks, ALB (perhaps in addition to NLB) Listeners and associating Target Groups and the failure of Terraform to account the Listener depending when modifying Target Groups.

**Comment by jbg** [comment]
Assuming everything is managed by Terraform (not other automation) then this works correctly, as long as you use the TF features that support this kind of dependency. Bear in mind that TF's graph doesn't know anything about "listeners" or "target groups", it just knows resources and dependencies between them. You need to use `create_before_destroy` because the order of operations needs to be:

1. New target group is created
2. Listener rule is updated to point to new target group
3. Old target group is deleted

Without `create_before_destroy` the order is this:

1. Old target group is deleted
2. New target group is created
3. Listener rule is updated to point to new target group

Which obviously fails at step 1 because AWS doesn't allow the deletion. This type of dependency is basically the reason for `create_before_destroy` existing.

You'll further need to use `name_prefix` because two target groups can't exist with the same name.

This isn't a bug, it's just still open because nobody really triages bugs on this repository (hence the 3.6k open issues).

**Comment by henryagbo** [comment]
@jbg  this is definitely a bug and the solution you proposed will certainly not scale when there is a name dependency on the created resource. This issue still happens even for environment I brought up using terraform 100%. It is either AWS allows edit of the target group that is attached to a load balancer, or they properly manage the dependency from their provider release

**Comment by jbg** [comment]
The provider is unable to do it, since the TF model builds the graph without any knowledge of what the resources represent. TF already allows you to design your resources in a way that avoids this issue (as explained above). If you have dependencies, managed outside TF, on TGs having specific names, that's a separate issue which nothing in TF can really save you from. If the name dependency is managed inside TF then it will work correctly.

**Comment by PeterBocan** [comment]
I am facing the same issue w/ ECS on Fargate. I find the notion of just chucking everything into the TF State a bit problematic. I think of the underlying provisioned platform as a separate abstraction layer from what ECS does to the services and tasks. (I think ECS is leaking too much information about the underlying architecture (notably subnets, and security groups) to a layer above, which should know nothing about where it is provisioned and what network rules it has attached to itself). 

I don't think of running the TF every time I change a parameter on a ECS Task Definition. I think that's an anti-pattern caused by the wrong ECS abstraction.

It is definitely biting me in the rear end and makes my life much harder than it should be.

**Comment by dga-nagra** [comment]
Hi everyone.

While the `name-prefix` + `lifecycle:create_before_destroy` combination works, it is indeed not optimal.
I assume that one way to fix that would be to be able to update existing target groups, which is not allowed by AWS.

A workaround would be that the terraform module:
- take track of the dependencies
- remove the relations
- destroy/re-recreate the resource
- redo all links
- (It may even need a temporary target group if a dependency cannot be left without target group)

This might not be to slow, but this is still a lot of work for something that does not concern everyone.

---
For the predictability, I assume that you can use a tag instead to identify the target group in your external tools.

---
Why is the name-prefix limited to 6 characters? If this is related to the number of characters allowed and the length of the suffix, then I think the module should be able to use a shorter suffix (to some extents) to allow a longer prefix and(/or) check potential clashes

**Comment by tetienne** [comment]
Indeed, a solution would be to combine name_prefix and create_before_destroy, but as said @dga-nagra name_prefix is limited to 6 chars.

So to bypass this, I used `uuid()`.

```hcl
resource "aws_lb_target_group" "foo" {
  name        = "${var.environment}-foo-${substr(uuid(), 0, 3)}"
  port        = local.container_port
  protocol    = "HTTP"
  target_type = "ip"
  vpc_id      = var.vpc_id
  lifecycle {
    create_before_destroy = true
  }
}
```

**Comment by jbg** [comment]
@tetienne that will recreate your target group every time you run Terraform, though...

**Comment by tetienne** [comment]
@jbg Indeed. So what about 

resource "aws_lb_target_group" "foo" {
  name        = "${var.environment}-foo-${substr(uuid(), 0, 3)}"
  port        = local.container_port
  protocol    = "HTTP"
  target_type = "ip"
  vpc_id      = var.vpc_id
  lifecycle {
    create_before_destroy = true
    ignore_changes        = [name]
  }
}

**Comment by jbg** [comment]
Fine if you don't want to change the name and are not interested in possible drift of the name. We just use a very short name prefix, and add a tag for the longer more descriptive name. Same pattern on many resources which have a short length constraint on the name.

**Comment by tetienne** [comment]
So finally, this solution allow to detect drift, and avoid conflicts when touching any attributes forcing a new creation:

```hcl
locals {
  tg_protocol    = "HTTP"
  tg_target_type = "ip"
}

resource "random_id" "suffix" {
  keepers = {
    protocol    = local.tg_protocol
    target_type = local.tg_target_type
    port        = local.container_port
    vpc_id      = var.vpc_id
  }

  byte_length = 2
}


resource "aws_lb_target_group" "foo" {
  name        = "${var.environment}-foo-${random_id.suffix.hex}"
  port        = local.container_port
  protocol    = local.tg_protocol
  target_type = local.tg_target_type
  vpc_id      = var.vpc_id
  health_check {
    path = "/health"
  }
  lifecycle {
    create_before_destroy = true
  }
}

```

**Comment by MartinEmrich** [comment]
Just saying: Any "solution" that involves changing the name of the target group with a randomized component is not a solution to at least some, I would argue even most of the subscribers here. Discussing the point of having fixed vs. randomized names IMHO also misses the point.

Doing it manually means detaching the group from the LB (i.e. removing the rule), deleting the non-migratable group, creating the new one with the same name, and then attaching the group to the LB again (i.e. creating the rule).

IMO TF should be capable of the same, and that's the heart of this Github issue.

**Comment by QuestionAndAnswer** [comment]
Solved this by using ```replace_triggered_by``` [link](https://developer.hashicorp.com/terraform/language/meta-arguments/lifecycle#replace_triggered_by). Shortly we just need to tell terraform to recreate whole tree of dependencies.

Having this,

```terraform
resource "aws_alb" "main" {
  name            = var.service_name
  subnets         = aws_subnet.public.*.id
  security_groups = [aws_security_group.alb.id]
}

resource "aws_alb_target_group" "app" {
  name        = var.service_name
  port        = 80
  protocol    = "HTTP"
  vpc_id      = aws_vpc.main.id
  target_type = "ip"

  health_check {
    healthy_threshold   = "3"
    interval            = "30"
    protocol            = "HTTP"
    matcher             = "200"
    timeout             = "3"
    path                = var.health_check_path
    unhealthy_threshold = "2"
  }
}

resource "aws_alb_listener" "http" {
  load_balancer_arn = aws_alb.main.id
  port              = 80
  protocol          = "HTTP"

  default_action {
    target_group_arn = aws_alb_target_group.app.id
    type             = "forward"
  }

  lifecycle {
    replace_triggered_by = [aws_alb_target_group.app]
  }
}

```

It forces terraform to first destory ```aws_alb_listener.http``` and only then ```aws_alb_target_group.app```, and recreate them in reverse, ```aws_alb_target_group.app``` and ```aws_alb_listener.http```

**Comment by Mmasson-01** [comment]
@QuestionAndAnswer  worked like a charm for me. Also tried all the other solutions proposed previously. This one seems to best to me.

**Comment by novekm** [comment]
I believe I have gotten it to work now. Unfortunately, I was unable to use an explicit dependency on my ALB with the ECS Cluster or Service. This is because to create the ECS Service (and thus the Cluster), the ALB and Target Group you define needs to be created **_before_** you attempt to create the cluster you reference them in. This means you cannot set the dependency this way to avoid the aforementioned issue.

After some more digging I found a two resources to get around this - [`aws_appautoscaling_target`](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/appautoscaling_target) and [`aws_appautoscaling_policy`](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/appautoscaling_policy). 

Essentially these resources give you some additional control over application scaling (in my use case ECS). This allows me to deregister the targets managed by ECS (when the `aws_appautoscaling_target` is destroyed, it sets the ECS container count back to 0. This allows me to successfully delete the listeners, as well as the ALB and target groups. They main thing is to have a lifecycle rule within the ECS Service resource that ignores changes to the `desired_count` :

```
  lifecycle {
    ignore_changes = [desired_count]  # Let Application Auto Scaling manage this
  }
```

To add to the complexity, I have my ALB, Listeners, and ECS Cluster in a parent module, while my Target Group, ECS Service, and ECS Task are in a child module. See the following example directory structure:

```
.
└── my-terraform-module/
    ├── modules/
    │   ├── p4-code-review
    │   ├── p4-auth
    │   └── p4-server
    └── main.tf
```

A similar code snippet to below resolved for me -  using a `null_resource` to pass the ARN of my ALB to the `aws_appautoscaling_target` and use it as a dependency (since you cannot use an upstream `depends_on`). If you're files are all within the same directory, you don't need the `null_resource` - you would just be able to use an explicit dependency with `depends_on` or by other means (such as referencing the ALB itself). The main thing is to ensure the Load Balancer gets created first by setting a dependency on the AWS Application Scaling resources.

```hcl
##########################################
# ECS Service
##########################################
resource "aws_ecs_service" "service" {
  name                = "${local.name_prefix}-service"
  cluster             = var.cluster_name != null ? data.aws_ecs_cluster.cluster[0].arn : aws_ecs_cluster.cluster[0].arn
  task_definition     = aws_ecs_task_definition.task_definition.arn
  desired_count       = var.desired_container_count
  launch_type         = "FARGATE"
  force_new_deployment = var.debug
  enable_execute_command = var.debug

  load_balancer {
    target_group_arn = aws_lb_target_group.alb_target_group.arn
    container_name   = var.container_name
    container_port   = var.container_port
  }

  network_configuration {
    subnets         = var.subnets
    security_groups = [aws_security_group.ecs_service.id]
  }

  lifecycle {
    ignore_changes = [desired_count]  # Let Application Auto Scaling manage this
  }

  depends_on = [
    aws_lb_target_group.alb_target_group
  ]

  tags = merge(var.tags,
    {
      Name = "${local.name_prefix}-service"
    }
  )
}

##########################################
# Null Resource
##########################################
# Used to set dependency on ALB from parent module
resource "null_resource" "parent_alb" {
  triggers = {
    shared_alb_arn = var.existing_application_load_balancer_arn
  }
}
##########################################
# Application Auto Scaling
##########################################
resource "aws_appautoscaling_target" "ecs_target" {
  max_capacity       = var.desired_container_count
  min_capacity       = 1
  resource_id        = "service/${var.project_prefix}-perforce-web-services-shared-cluster/${aws_ecs_service.service.name}"
  scalable_dimension = "ecs:service:DesiredCount"
  service_namespace  = "ecs"

  depends_on = [
    aws_ecs_service.service
  ]
}

resource "aws_appautoscaling_policy" "scale_up" {
  name               = "${local.name_prefix}-scale-up"
  resource_id        = aws_appautoscaling_target.ecs_target.resource_id
  scalable_dimension = aws_appautoscaling_target.ecs_target.scalable_dimension
  service_namespace  = aws_appautoscaling_target.ecs_target.service_namespace

  step_scaling_policy_configuration {
    adjustment_type         = "ExactCapacity"
    cooldown               = 60
    metric_aggregation_type = "Average"

    step_adjustment {
      metric_interval_lower_bound = 0
      scaling_adjustment          = var.desired_container_count
    }
  }

  depends_on = [
    aws_lb_target_group.alb_target_group,
    aws_ecs_service.service,
    aws_appautoscaling_target.ecs_target
  ]
}

##########################################
# Target Group for shared ALB
##########################################
resource "aws_lb_target_group" "alb_target_group" {
  name        = "${local.name_prefix}-tg"
  port        = var.container_port
  protocol    = "HTTP"
  target_type = "ip"
  vpc_id      = var.vpc_id

  health_check {
    path                = "/"
    healthy_threshold   = 2
    unhealthy_threshold = 2
    timeout             = 10
    interval            = 30
    matcher             = "200"
  }

  tags = merge(var.tags,
    {
      Name = "${local.name_prefix}-tg"
    }
  )
}
```

With the above I can successfully create and destroy without issue:

```hcl
module.terraform-aws-perforce.aws_ecs_cluster.perforce_web_services_cluster[0]: Destruction complete after 0s
aws_vpc.perforce_vpc: Destruction complete after 0s

Destroy complete! Resources: 121 destroyed.
```

Hope this helps others.
